File: profile/httpclient/codec.go
package httpclient

import (
	"context"

	// Packages
	client "github.com/mutablelogic/go-client"
	schema "github.com/mutablelogic/go-media/profile/schema"
	types "github.com/mutablelogic/go-server/pkg/types"
)

///////////////////////////////////////////////////////////////////////////////
// PUBLIC METHODS

func (c *Client) ListCodecs(ctx context.Context, req schema.CodecListRequest) (*schema.CodecList, error) {
	// Perform request
	var response schema.CodecList
	if err := c.DoWithContext(ctx, nil, &response, client.OptPath("codec"), client.OptQuery(req.Query())); err != nil {
		return nil, err
	}

	// Return the response
	return types.Ptr(response), nil
}

func (c *Client) GetCodec(ctx context.Context, name string) (*schema.Codec, error) {
	// Perform request
	var response schema.Codec
	if err := c.DoWithContext(ctx, nil, &response, client.OptPath("codec", name)); err != nil {
		return nil, err
	}

	// Return the response
	return types.Ptr(response), nil
}

File: ts/httpclient/codec.ts
// Mirrors profile/httpclient/codec.go

import type { Client } from "./client.js";
import type { Codec, CodecFilterType, CodecList, CodecListRequest } from "../schema/codec.js";

// The server's query-string binder (httprequest.Query) reads "type" as a raw
// integer into CodecType's underlying int, rather than through the string-based
// JSON (un)marshaling CodecType otherwise uses. These ordinals are libavutil's
// AVMediaType C enum values, which are stable across ffmpeg versions.
const AVMEDIA_TYPE: Record<CodecFilterType, number> = {
  video: 0,
  audio: 1,
  subtitle: 3,
};

function query(req: CodecListRequest): URLSearchParams {
  const params = new URLSearchParams();
  if (req.type) {
    params.set("type", String(AVMEDIA_TYPE[req.type]));
  }
  if (req.offset) {
    params.set("offset", String(req.offset));
  }
  if (req.limit !== undefined) {
    params.set("limit", String(req.limit));
  }
  return params;
}

export function listCodecs(client: Client, req: CodecListRequest = {}): Promise<CodecList> {
  return client.request<CodecList>(["codec"], query(req));
}

export function getCodec(client: Client, name: string): Promise<Codec> {
  return client.request<Codec>(["codec", name]);
}
